README.md
# IvcvEx

<!-- ![GitHub release (latest by date)](https://img.shields.io/github/v/release/bulld0zer/ivcv_ex) -->
![Hex.pm](https://img.shields.io/hexpm/v/ivcv_ex)
![Hex.pm](https://img.shields.io/hexpm/dt/ivcv_ex)
[![Docs](https://img.shields.io/badge/hex-docs-blue)](https://wois.hexdocs.pm/ivcv_ex/)

IvcvEx is a client library for [IVCV](https://ivcv.eu/), so you can request video analysis and later fetch the results.

You only need to set your authentication api key in the config file and enjoy the feature of IVCV withou having to configure http calls, ensure urls or dealing with tricky configurations. Use your time in the business logic of your app.


-----
## Installation

The package can be installed by adding `ivcv_ex` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ivcv_ex, "~> 0.1.0"}
  ]
end
```
-----
## Configurations

* config.exs

```elixir
config :ivcv_ex,
  http_client: HTTPoison,
  auth_key: "[your_authentication_key]",
  base_url: "https://api.ivcv.eu"
```

-----

## License
Licensed under [MIT license](LICENSE)


-----
## PS
If you found this library useful, dont forget to star it (on github) =)

![GitHub stars](https://img.shields.io/github/stars/wois-org/ivcv_ex?style=social)



CHANGELOG.md
# Changelog
<!-- next-header -->

## [Unreleased] - ReleaseDate

## [0.24.0] - 2023-9-28

## [0.23.1-rc.0] - 2023-9-27
### Changed
* updated ivcv result structure 

## [0.23.0] - 2023-9-20
### Added
* extra field to the structure of IVCV analysis result
    - eye_contact
    - confidence
    - pitch_average
    - pitch_std
    - words_per_min
    - num_of_pauses
    - pitch_mean_level
    - pitch_std_level

## [0.22.0] - 2023-4-26
### Added
* stub for video processing with status PROCESSING 

## [0.21.0] - 2023-4-25
### Added
* stub for video processing failed 

## [0.20.0] - 2023-4-25
### Fix
* parse video_processing_by_url result

## [0.19.0] - 2023-4-24
### Fix
* request body parameters
* increased receive timeout to 15 seconds

## [0.18.0] - 2023-4-21
### Fix
* body request for videoUrl on video processing

## [0.17.0] - 2023-4-21
### Fix
* temporary remove spec to fix dialyzer in child projects

## [0.16.0] - 2023-4-19
### Fix
* log message on ivcv error request

## [0.15.0] - 2023-4-19
### Fix
* description and readme

## [0.14.0] - 2023-4-19
### Fix
* github actions configuration, set remote origin url properly

## [0.13.0] - 2023-4-19
### Fix
* github actions configuration, set remote origin url properly

## [0.12.0] - 2023-4-19
### Fix
* github actions configuration, set remote origin url properly

## [0.11.0] - 2023-4-19
### Fix
* version release configuration, no need to maintain develop branch
* github actions configuration

## [0.10.0] - 2023-4-19
### Fix
* Error parsing from IVCV

## [0.9.0] - 2023-4-19
### Added
* stub for video analysis request error

## [0.8.1-rc.0] - 2023-4-17
### Added
* get video analisys result endpoint

## [0.8.0] - 2023-4-5
### Dependencies
* version_release 0.5.1 => 0.5.2

## [0.7.2] - 2023-4-5

## [0.7.1] - 2023-4-5

## [0.7.0] - 2023-4-5
### Fix
* enabled hex force publish config for version_release

## [0.6.0] - 2023-4-5
### Dependencies
* version_release 0.5.0 => 0.5.1

## [0.5.4] - 2023-3-31

## [0.5.3] - 2023-3-31

## [0.5.2] - 2023-3-31

## [0.5.1] - 2023-3-31

## [0.5.0] - 2023-3-31

## [0.4.2] - 2023-3-31

## [0.4.1] - 2023-3-31

## [0.4.0] - 2023-3-31

## [0.3.0] - 2023-3-31

## [0.2.0] - 2023-3-31
### Added
* Initial setup

<!-- next-url -->
[Unreleased]: https://github.com/wois-org/ivcv_ex/compare/v0.24.0...HEAD
[0.24.0]: https://github.com/wois-org/ivcv_ex/compare/v0.23.1-rc.0...v0.24.0
[0.23.1-rc.0]: https://github.com/wois-org/ivcv_ex/compare/v0.23.0...v0.23.1-rc.0
[0.23.0]: https://github.com/wois-org/ivcv_ex/compare/v0.22.0...v0.23.0
[0.22.0]: https://github.com/wois-org/ivcv_ex/compare/v0.21.0...v0.22.0
[0.21.0]: https://github.com/wois-org/ivcv_ex/compare/v0.20.0...v0.21.0
[0.20.0]: https://github.com/wois-org/ivcv_ex/compare/v0.19.0...v0.20.0
[0.19.0]: https://github.com/wois-org/ivcv_ex/compare/v0.18.0...v0.19.0
[0.18.0]: https://github.com/wois-org/ivcv_ex/compare/v0.17.0...v0.18.0
[0.17.0]: https://github.com/wois-org/ivcv_ex/compare/v0.16.0...v0.17.0
[0.16.0]: https://github.com/wois-org/ivcv_ex/compare/v0.15.0...v0.16.0
[0.15.0]: https://github.com/wois-org/ivcv_ex/compare/v0.14.0...v0.15.0
[0.14.0]: https://github.com/wois-org/ivcv_ex/compare/v0.13.0...v0.14.0
[0.13.0]: https://github.com/wois-org/ivcv_ex/compare/v0.12.0...v0.13.0
[0.12.0]: https://github.com/wois-org/ivcv_ex/compare/v0.11.0...v0.12.0
[0.11.0]: https://github.com/wois-org/ivcv_ex/compare/v0.10.0...v0.11.0
[0.10.0]: https://github.com/wois-org/ivcv_ex/compare/v0.9.0...v0.10.0
[0.9.0]: https://github.com/wois-org/ivcv_ex/compare/v0.8.1-rc.0...v0.9.0
[0.8.1-rc.0]: https://github.com/wois-org/papelillo/compare/v0.8.0...v0.8.1-rc.0
[0.8.0]: https://github.com/wois-org/papelillo/compare/v0.7.2...v0.8.0
[0.7.2]: https://github.com/wois-org/papelillo/compare/v0.7.1...v0.7.2
[0.7.1]: https://github.com/wois-org/papelillo/compare/v0.7.0...v0.7.1
[0.7.0]: https://github.com/wois-org/papelillo/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/wois-org/papelillo/compare/v0.5.4...v0.6.0
[0.5.4]: https://github.com/wois-org/papelillo/compare/v0.5.3...v0.5.4
[0.5.3]: https://github.com/wois-org/papelillo/compare/v0.5.2...v0.5.3
[0.5.2]: https://github.com/wois-org/papelillo/compare/v0.5.1...v0.5.2
[0.5.1]: https://github.com/wois-org/papelillo/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/wois-org/papelillo/compare/v0.4.2...v0.5.0
[0.4.2]: https://github.com/wois-org/papelillo/compare/v0.4.1...v0.4.2
[0.4.1]: https://github.com/wois-org/papelillo/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/wois-org/papelillo/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/wois-org/papelillo/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/wois-org/ivcv_ex/compare/v0.1.0...v0.2.0


config/test.exs
import Config

config :ivcv_ex,
  http_client: HTTPoisonMock,
  env: "test"

import_config "version_release.exs"


test/ivcv_ex_test.exs
defmodule IvcvExTest do
  use ExUnit.Case
  import Mox

  alias IvcvEx

  setup :set_mox_global
  setup :verify_on_exit!

  describe "request video processing" do
    test "with correct params" do
      expect(HTTPoisonMock, :post, 1, fn _url, _body, _headers, _opts ->
        IvcvEx.Stub.process_video_ok()
      end)

      assert {:ok, _request_id} = IvcvEx.process_video_by_url("https://video_url")
    end
  end

  describe "get video analysis result" do
    test "with correct params" do
      expect(HTTPoisonMock, :post, 1, fn _url, _body, _headers, _opts ->
        IvcvEx.Stub.get_analysis_result_ok()
      end)

      assert {:ok,
              %IvcvEx.Result{
                result_id: "abc123",
                status: "FINISHED",
                impression: %IvcvEx.Result.Impression{
                  neuroticism: 70,
                  extraversion: 64,
                  openness: 70,
                  conscientiousness: 69,
                  agreeableness: 69
                },
                emotions: %IvcvEx.Result.Emotions{intensity: 46, positivity: 11},
                sentiment: 61,
                eye_contact: 0.67,
                confidence: 62,
                pitch_average: 109.0,
                pitch_average_level: "low",
                pitch_std_level: "medium",
                words_per_min: 44.0,
                num_of_pauses: 8,
                pauses_per_min: 2.0
              }} = IvcvEx.process_video_by_url("https://video_url")
    end
  end
end


mix.exs
defmodule IvcvEx.MixProject do
  use Mix.Project

  def project do
    [
      app: :ivcv_ex,
      version: "0.24.1-alpha.0",
      elixir: "~> 1.12",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      name: "IvcvEx",
      description: description(),
      package: package(),
      test_coverage: [tool: ExCoveralls],
      docs: [
        main: "readme",
        extras: ["README.md", "CHANGELOG.md", "LICENSE"]
      ],
      source_url: "https://github.com/wois-org/ivcv_ex"
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp description() do
    "IVCV Client Library"
  end

  defp package() do
    [
      name: "ivcv_ex",
      files: ~w(lib .formatter.exs mix.exs README* LICENSE CHANGELOG*),
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/wois-org/ivcv_ex"}
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:httpoison, "~> 2.1"},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:credo, "~> 1.7", only: [:dev, :test], runtime: false},
      {:excoveralls, "0.16.1", only: :test},
      {:version_release, "0.5.2", only: [:test, :dev], runtime: false},
      {:mox, "~> 1.0", only: :test}
    ]
  end
end


lib/ivcv_ex.ex
defmodule IvcvEx do
  require Logger

  alias IvcvEx.Result

  @moduledoc """
  Documentation for `IvcvEx`.
  """

  @doc """
  Process video by url.

  Returns `{:ok, result_id}` if video was accepted by IVCV servers, otherwise `{:error, reason}` is returned.

  ## Parameters

    - video_url: String that represents the video url.

  ## Examples

      IvcvEx.process_video_by_url("https://example.com/video.mp4")

  ## Reference
    https://docs.ivcv.eu/result_for_video.html#blueprints.direct_blueprints.process_video_by_url
  """
  def process_video_by_url(video_url) when is_binary(video_url) do
    path = "/video"
    auth_api_key = Application.get_env(:ivcv_ex, :auth_key)
    http_client = Application.get_env(:ivcv_ex, :http_client)

    base_url = Application.get_env(:ivcv_ex, :base_url)

    url = base_url <> path

    body =
      %{"videoUrl" => video_url}
      |> Jason.encode!()

    headers = [
      {"Content-Type", "application/json"},
      {"Authorization", auth_api_key}
    ]

    http_client.post(url, body, headers, recv_timeout: 15_000)
    |> parse_response()
  end

  @doc """
  Get the result of a video analysis. Result include multiple items from the analysis.

  Returns `{:ok, Result.t()}` and Result contains:
    emotions (object) - Average intensity and positivity during the video. Integer values between 0 to 100.

    impression (object) - First impression analysis with Big Five metrics. Integer values between 0 to 100.

    sentiment (float) - Sentiment analysis result. Integer value between 0 to 100.

    status (string) - Current status of the processing. One of FINISHED, PROCESSING, FAILED.

    resultId (string) - Id of the result.

  ## Parameters

    - result_id: Id of the result. This is obtained when the video is sent initially.

  ## Examples

      IvcvEx.get_analysis_result("result_id")

  ## Reference
    https://docs.ivcv.eu/result_for_video.html#blueprints.direct_blueprints.analysis_result
  """
  def get_analysis_result(result_id) do
    path = "/result"
    auth_api_key = Application.get_env(:ivcv_ex, :auth_key)
    http_client = Application.get_env(:ivcv_ex, :http_client)

    base_url = Application.get_env(:ivcv_ex, :base_url)

    url = base_url <> path

    body =
      %{"resultId" => result_id}
      |> Jason.encode!()

    headers = [
      {"Content-Type", "application/json"},
      {"Authorization", auth_api_key}
    ]

    http_client.post(url, body, headers, recv_timeout: 15_000)
    |> parse_response()
  end

  defp parse_response({:ok, %HTTPoison.Response{status_code: 200, body: body}}) do
    body
    |> Jason.decode()
    |> case do
      {:ok, response} ->
        Result.parse(response)

      {:error, reason} ->
        Logger.error(
          "#{inspect(__MODULE__)} something went wrong, could not decode response: #{inspect(reason, pretty: true)}"
        )

        {:error, reason}
    end
  end

  defp parse_response({:ok, %HTTPoison.Response{status_code: status_code} = resp}) do
    Logger.error("IVCV responded with code: #{status_code}", response: resp)
    {:error, "provider_response_status_code_#{status_code}"}
  end

  defp parse_response({:ok, %HTTPoison.Response{status_code: _, body: body}}) do
    Logger.error(
      "#{inspect(__MODULE__)} Internal server error, returned: #{inspect(body, pretty: true)}"
    )

    {:error, "video analysis failed"}
  end

  defp parse_response({:error, %HTTPoison.Error{reason: reason}}) do
    {:error, reason}
  end
end


lib/ivcv/result.ex
defmodule IvcvEx.Result.Impression do
  @moduledoc false
  @derive {Jason.Encoder, except: []}
  @type t :: %__MODULE__{
          neuroticism: integer(),
          extraversion: integer(),
          openness: integer(),
          conscientiousness: integer(),
          agreeableness: integer()
        }
  @enforce_keys [:neuroticism, :extraversion, :openness, :conscientiousness, :agreeableness]

  alias IvcvEx.Result.Impression

  defstruct [
    :neuroticism,
    :extraversion,
    :openness,
    :conscientiousness,
    :agreeableness
  ]

  def parse(%{
        "neuroticism" => neuroticism,
        "extraversion" => extraversion,
        "openness" => openness,
        "conscientiousness" => conscientiousness,
        "agreeableness" => agreeableness
      }) do
    %Impression{
      neuroticism: neuroticism,
      extraversion: extraversion,
      openness: openness,
      conscientiousness: conscientiousness,
      agreeableness: agreeableness
    }
  end
end

defmodule IvcvEx.Result.Emotions do
  @moduledoc false
  @derive {Jason.Encoder, except: []}
  @type t :: %__MODULE__{
          intensity: integer(),
          positivity: integer()
        }
  @enforce_keys [:intensity, :positivity]

  alias IvcvEx.Result.Emotions

  defstruct [
    :intensity,
    :positivity
  ]

  def parse(%{"intensity" => intensity, "positivity" => positivity}) do
    %Emotions{intensity: intensity, positivity: positivity}
  end
end

defmodule IvcvEx.Result do
  @moduledoc """
  Struct to hold the data analysis result
    emotions (Emotions.t()) - Average intensity and positivity during the video. Integer values between 0 to 100.

    impression (Impression.t()) - First impression analysis with Big Five metrics. Integer values between 0 to 100.

    sentiment (integer) - Sentiment analysis result. Integer value between 0 to 100.

    eyeContact (float) - Eye contact ratio of the person during the video. Float value between 0 to 1.

    confidence (int) - Confidence level of the person. Integer value between 0 to 100.

    pitchAverage (float) - Average pitch of the speech. Integer value between 75 to 600.

    pitchStd (float) - Standard deviation of the pitch. Integer value between 0 to 600.

    pitchAverageLevel (string) - Level of pitch average. One of “low”, “medium”, “high”.

    pitchStdLevel (string) - Level of pitch std. One of “low”, “medium”, “high”.

    wordsPerMin (float) - Words per minute spoken by the person.

    numOfPauses (int) - Number of pauses during the speech. Minimum pause duration is 0.3 seconds.

    pausesPerMin (float) - Pauses per minute during the speech.

    status (string) - Current status of the processing. One of FINISHED, PROCESSING, FAILED.

    resultId (string) - Id of the result.

  """
  require Logger

  alias IvcvEx.Result
  alias IvcvEx.Result.Emotions
  alias IvcvEx.Result.Impression

  @derive {Jason.Encoder, except: []}
  @type t :: %__MODULE__{
          result_id: binary,
          status: String.t(),
          impression: Impression,
          emotions: Emotions,
          sentiment: integer(),
          eye_contact: float(),
          confidence: integer(),
          pitch_average: float(),
          pitch_std: float(),
          pitch_average_level: String.t(),
          pitch_std_level: String.t(),
          words_per_min: float(),
          num_of_pauses: integer(),
          pauses_per_min: float()
        }
  defstruct [
    :result_id,
    :status,
    :impression,
    :emotions,
    :sentiment,
    :eye_contact,
    :confidence,
    :pitch_average,
    :pitch_std,
    :pitch_average_level,
    :pitch_std_level,
    :words_per_min,
    :num_of_pauses,
    :pauses_per_min
  ]

  def parse(%{
        "resultId" => result_id,
        "status" => status,
        "impression" => impression,
        "emotions" => emotions,
        "sentiment" => sentiment,
        "eyeContact" => eye_contact,
        "confidence" => confidence,
        "pitchAverage" => pitch_average,
        "pitchStd" => pitch_std,
        "pitchAverageLevel" => pitch_mean_level,
        "pitchStdLevel" => pitch_std_level,
        "wordsPerMin" => words_per_min,
        "numOfPauses" => num_of_pauses,
        "pausesPerMin" => pauses_per_min
      }) do
    {:ok,
     %Result{
       result_id: result_id,
       status: status,
       impression: Impression.parse(impression),
       emotions: Emotions.parse(emotions),
       sentiment: sentiment,
       eye_contact: eye_contact,
       confidence: confidence,
       pitch_average: pitch_average,
       pitch_std: pitch_std,
       pitch_average_level: pitch_mean_level,
       pitch_std_level: pitch_std_level,
       words_per_min: words_per_min,
       num_of_pauses: num_of_pauses,
       pauses_per_min: pauses_per_min
     }}
  end

  def parse(%{
        "status" => status,
        "resultId" => result_id
      }) do
    {:ok,
     %{
       status: status,
       result_id: result_id
     }}
  end

  def parse(%{
        "resultId" => result_id
      }) do
    {:ok,
     %{
       result_id: result_id
     }}
  end

  def parse(input) do
    Logger.error(
      "#{inspect(__MODULE__)} something went wrong, could not parse input: #{inspect(input, pretty: true)}"
    )

    {:error, input}
  end
end


config/version_release.exs
import Config

replacements = [
  %{
    file: "CHANGELOG.md",
    type: :changelog,
    patterns: [
      %{search: "Unreleased", replace: "{{version}}", type: :unreleased},
      %{search: "...HEAD", replace: "...{{tag_name}}", global: false},
      %{search: "ReleaseDate", replace: "{{date}}"},
      %{
        search: "<!-- next-header -->",
        replace: "<!-- next-header -->\n\n## [Unreleased] - ReleaseDate",
        global: false
      },
      %{
        search: "<!-- next-url -->",
        replace:
          "<!-- next-url -->\n[Unreleased]: https://github.com/wois-org/ivcv_ex/compare/{{tag_name}}...HEAD",
        global: false
      }
    ]
  }
]

config :version_release,
  tag_prefix: "v",
  hex_publish: true,
  hex_force_publish: true,
  changelog: %{
    creation: :manual,
    minor_patterns: ["added", "changed", "fixed", "fix"],
    major_patterns: ["breaking"],
    replacements: replacements,
    pre_release_replacements: replacements
  },
  commit_message: "[skip ci][version_release] {{message}}",
  dev_version: true


lib/ivcv/stub.ex
defmodule IvcvEx.Stub do
  @moduledoc false
  def process_video_ok() do
    {:ok,
     %HTTPoison.Response{status_code: 200, body: %{"resultId" => "abc123"} |> Jason.encode!()}}
  end

  def get_analysis_result_ok() do
    {:ok,
     %HTTPoison.Response{
       status_code: 200,
       body:
         %{
           "resultId" => "abc123",
           "status" => "FINISHED",
           "impression" => %{
             "neuroticism" => 70,
             "extraversion" => 64,
             "openness" => 70,
             "conscientiousness" => 69,
             "agreeableness" => 69
           },
           "emotions" => %{
             "intensity" => 46,
             "positivity" => 11
           },
           "sentiment" => 61,
           "eyeContact" => 0.67,
           "confidence" => 62,
           "pitchAverage" => 109.0,
           "pitchStd" => 27.0,
           "pitchAverageLevel" => "low",
           "pitchStdLevel" => "medium",
           "wordsPerMin" => 44.0,
           "numOfPauses" => 8,
           "pausesPerMin" => 2.0
         }
         |> Jason.encode!()
     }}
  end

  def get_analysis_result_processing() do
    {:ok,
     %HTTPoison.Response{
       status_code: 200,
       body:
         %{
           "resultId" => "abc123",
           "status" => "PROCESSING"
         }
         |> Jason.encode!()
     }}
  end

  def get_analysis_result_failed() do
    {:ok,
     %HTTPoison.Response{
       status_code: 200,
       body:
         %{
           "resultId" => "abc123",
           "status" => "FAILED"
         }
         |> Jason.encode!()
     }}
  end

  def get_analysis_result_error() do
    {:ok,
     %HTTPoison.Response{
       status_code: 500,
       body:
         %{
           "resultId" => "abc123",
           "status" => "FAILED"
         }
         |> Jason.encode!()
     }}
  end
end
